File: 257_Binary_Tree_Paths.java
/*
https://leetcode.com/problems/binary-tree-paths/
Given a binary tree, return all root-to-leaf paths.

Top-down recursively add node.val to path and recurse if node is not a leaf, or store result for leaves.
Alternatively, bottom-up find all paths from left and right nodes to leaves a prepend current node.
Both can be converted to use StringBuilder but leaf node values must be removed from path after toString() in order
to revert to original path.
Time - O(n**2), for every node visited extend existing path
Space - O(n)
*/

public class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> paths = new ArrayList<String>();       // LinkedList also works
        helper(root, "", paths);
        return paths;
    }

    public void helper(TreeNode node, String path, List<String> paths) {
        if (node == null)
            return;
        if (node.left == null && node.right == null)        // leaf node, add to result
            paths.add(path + node.val);
        helper(node.left, path + node.val + "->", paths);
        helper(node.right, path + node.val + "->", paths);
    }
}


public class Solution2 {
    public List<String> binaryTreePaths(TreeNode root) {

        List<String> paths = new ArrayList<String>();

        if (root == null)
            return paths;

        if (root.left == null && root.right == null) {
            paths.add(Integer.toString(root.val));
            return paths;
        }

        List<String> left_paths = binaryTreePaths(root.left);
        List<String> right_paths = binaryTreePaths(root.right);

        for (String p : left_paths)
            paths.add(root.val + "->" + p);
        for (String p : right_paths)
            paths.add(root.val + "->" + p);

        return paths;
    }
}

File: python_1_to_1000/257_Binary_Tree_Paths.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/binary-tree-paths/
# Given a binary tree, return all root-to-leaf paths.

# Recursive dfs. If leaf, add path to result. Else add node value to partial and recurse.
# Time - O(n**2)
# Space - O(n)

class Solution(object):
    def binaryTreePaths(self, root):
        """
        :type root: TreeNode
        :rtype: List[str]
        """

        def helper(node, partial):          # partial is exiting path from root
            if not node:
                return
            partial.append(str(node.val))
            if not node.left and not node.right:
                paths.append("->".join(partial))
                return
            helper(node.left, partial[:])
            helper(node.right, partial)

        paths = []
        helper(root, [])
        return paths
